File: slk_bot/slk/engine.py
"""Layer 2 — XYZ execution engine (confirmation-entry mode).

Replays the trailing ``setup_window`` of entry-timeframe candles against
point-in-time storyline snapshots and walks each direction through:

    MAP      an armed origin key level on the correct side of price
    TOUCH    price trades into the origin zone
    SWEEP    counter-side internal liquidity is swept inside the zone
             (wick beyond a resting swing pool, close back inside)
    SHIFT    entry-timeframe BOS: close beyond the pullback structure
    RETEST   return to the V-level/origin zone  → ALERT
    INVALID  close beyond the sweep extreme / key level / HTF story flip
    EXPIRED  a stage timed out

Confirmation-entry mode is the only mode (direct key-level entries stay
disabled, per the research). Every transition is emitted as an auditable
Event; alerts carry the full persistence record. The engine is stateless
across scans — idempotency comes from event/alerts dedupe keys in the DB.
"""
from __future__ import annotations

import hashlib
import logging
from bisect import bisect_right
from datetime import timedelta

from ..models import Candle, Direction
from . import features as F
from .storyline import MAP_TF_SECONDS
from .types import Alert, Event, Storyline, Setup

log = logging.getLogger(__name__)

PARAM_VERSION = "slk-r2.0"


def _setup_id(pair: str, entry_tf: str, d: Direction, level, story_asof) -> str:
    raw = (
        f"{pair}|{entry_tf}|{d.value}|{level.kind}|{level.origin_price:.6f}|"
        f"{level.origin_time.isoformat()}|{story_asof.isoformat()}"
    )
    return hashlib.sha1(raw.encode()).hexdigest()[:12]


def scan_entry(
    *,
    pair: str,
    entry_tf: str,
    tf_seconds: int,
    candles: list[Candle],
    snaps: list[tuple, list],
    cfg,
    mode: str = "paper",
) -> tuple[list[Alert], list[Event]]:
    """Replay entry candles and return (alerts, events). ``snaps`` is the
    output of ``storyline_series`` — [(h4_open_time, Storyline)]."""
    alerts: list[Alert] = []
    events: list[Event] = []
    min_bars = cfg.pivot_left + cfg.pivot_right + 6
    if len(candles) < min_bars:
        return alerts, events

    # snapshot validity starts when that H4 candle has closed
    valid_from = sorted(t + timedelta(seconds=MAP_TF_SECONDS) for t, _ in snaps)
    story_by_key = {t + timedelta(seconds=MAP_TF_SECONDS): s for t, s in snaps}

    def story_at(close_time) -> Storyline | None:
        i = bisect_right(valid_from, close_time) - 1
        return story_by_key[valid_from[i]] if i >= 0 else None

    atr_e = F.atr(candles, cfg.atr_period)
    highs, lows = F.find_swings(candles, cfg.pivot_left, cfg.pivot_right)
    conf = cfg.pivot_right  # a pool is only usable once confirmed
    start = max(0, len(candles) - cfg.setup_window)
    active: dict[Direction, Setup | None] = {Direction.LONG: None, Direction.SHORT: None}

    def emit(s: Setup, state: str, c: Candle, reason: str, price=None) -> None:
        events.append(Event(s.setup_id, pair, state, c.time, reason, price))
        log.info("%-7s %s %s %s — %s", state, pair, entry_tf, s.setup_id, reason)

    def kill(s: Setup, state: str, c: Candle, reason: str) -> None:
        emit(s, state, c, reason, c.close)
        active[s.direction] = None

    for i in range(start, len(candles)):
        c = candles[i]
        close_time = c.time + timedelta(seconds=tf_seconds)
        story = story_at(close_time)

        for d in (Direction.SHORT, Direction.LONG):
            is_short = d is Direction.SHORT
            s = active[d]

            # ---- arming (MAP) ------------------------------------------
            if s is None and story and story.valid and story.direction is d and story.origin:
                lv = story.origin
                on_side = c.close < lv.zone_lo if is_short else c.close > lv.zone_hi
                if on_side:
                    sid = _setup_id(pair, entry_tf, d, lv, story.asof)
                    s = Setup(
                        setup_id=sid, direction=d, level=lv,
                        map_index=i, map_time=c.time,
                        environment=story.environment, phase=story.phase,
                        htf_alignment=story.htf_alignment,
                        draw_on_liquidity=story.draw_on_liquidity,
                        nearest_external_target=story.nearest_external_target,
                        internal_pools=list(story.internal_pools),
                        external_pools=list(story.external_pools),
                        imbalances=list(story.imbalances),
                    )
                    active[d] = s
                    emit(s, "MAP", c,
                         f"{lv.kind}-level {lv.origin_price:g} armed "
                         f"({story.environment}/{story.phase})",
                         lv.origin_price)
            if s is None:
                continue

            z = s.level

            # ---- HTF story flip ----------------------------------------
            if story is not None and story.valid and story.direction is not d:
                kill(s, "INVALID", c, "HTF storyline invalidation")
                continue

            broke_level = (
                c.close > z.zone_hi + cfg.flip_margin_atr * atr_e
                if is_short
                else c.close < z.zone_lo - cfg.flip_margin_atr * atr_e
            )

            # ---- MAP → TOUCH -------------------------------------------
            if s.state == "MAP":
                if broke_level:
                    kill(s, "INVALID", c, "close beyond key level (level may flip)")
                    continue
                touched = c.high >= z.zone_lo if is_short else c.low <= z.zone_hi
                if touched:
                    s.state = "TOUCH"
                    s.touch_index, s.touch_time = i, c.time
                    emit(s, "TOUCH", c, "price entered the origin zone", c.close)
                elif i - s.map_index > cfg.touch_window:
                    kill(s, "EXPIRED", c, "level not reached in time")
                    continue

            # ---- TOUCH → SWEEP -----------------------------------------
            if s.state == "TOUCH":
                if broke_level:
                    kill(s, "INVALID", c, "close beyond key level without a sweep")
                    continue
                pools = highs if is_short else lows
                swept = [
                    sw for sw in pools
                    if sw.index <= s.touch_index and sw.index + conf <= i
                    and (
                        (c.high > sw.price and c.close < sw.price)
                        if is_short
                        else (c.low < sw.price and c.close > sw.price)
                    )
                    and (c.high >= z.zone_lo if is_short else c.low <= z.zone_hi)
                ]
                if swept:
                    pick = max(swept, key=lambda w: w.price) if is_short else min(
                        swept, key=lambda w: w.price
                    )
                    s.swept_pool_index, s.swept_pool_price = pick.index, pick.price
                    s.sweep_index, s.sweep_time = i, c.time
                    s.extreme = c.high if is_short else c.low
                    refs = [
                        rw for rw in (lows if is_short else highs)
                        if pick.index < rw.index <= i and rw.index + conf <= i
                    ]
                    if refs:
                        s.ref_price = refs[-1].price
                    else:
                        seg = candles[s.touch_index : i + 1]
                        s.ref_price = (
                            min(x.low for x in seg) if is_short
                            else max(x.high for x in seg)
                        )
                    s.state = "SHIFT"
                    emit(s, "SWEEP", c,
                         f"swept {'buyside' if is_short else 'sellside'} internal "
                         f"liquidity @ {pick.price:g}", pick.price)
                elif i - s.touch_index > cfg.sweep_window:
                    kill(s, "EXPIRED", c, "no liquidity sweep after the touch")
                    continue

            # ---- SWEEP → SHIFT (BOS) -----------------------------------
            if s.state == "SHIFT":
                if i > s.sweep_index:
                    prev = candles[i - 1]
                    s.extreme = (
                        max(s.extreme, prev.high) if is_short
                        else min(s.extreme, prev.low)
                    )
                violated = c.close > s.extreme if is_short else c.close < s.extreme
                if violated:
                    kill(s, "INVALID", c, "close beyond sweep extreme")
                    continue
                bos = c.close < s.ref_price if is_short else c.close > s.ref_price
                if bos:
                    s.inv_level = (
                        max(s.extreme, c.high) if is_short
                        else min(s.extreme, c.low)
                    )
                    s.bos_index, s.bos_time = i, c.time
                    s.state = "RETEST"
                    emit(s, "SHIFT", c,
                         f"BOS through pullback structure {s.ref_price:g}", c.close)
                elif i - s.sweep_index > cfg.bos_window:
                    kill(s, "EXPIRED", c, "no BOS after the sweep")
                    continue

            # ---- SHIFT → RETEST → ALERT --------------------------------
            if s.state == "RETEST":
                violated = (
                    c.close > s.inv_level if is_short else c.close < s.inv_level
                )
                if violated:
                    kill(s, "INVALID", c, "close beyond invalidation level")
                    continue
                if i > s.bos_index:  # retest must be a candle after the BOS
                    left = c.close < z.zone_lo if is_short else c.close > z.zone_hi
                    if left:
                        s.left_zone = True
                    tol = cfg.retest_tolerance_atr * atr_e
                    returns = (
                        c.high >= z.zone_lo - tol if is_short
                        else c.low <= z.zone_hi + tol
                    )
                    if s.left_zone and returns:
                        # opposing liquidity must remain standing for a
                        # reversal-style setup: the draw pool must be
                        # untouched since the setup was armed
                        if s.draw_on_liquidity is not None:
                            seg = candles[s.map_index : i + 1]
                            standing = (
                                all(x.low > s.draw_on_liquidity for x in seg)
                                if is_short
                                else all(x.high < s.draw_on_liquidity for x in seg)
                            )
                        else:
                            standing = False
                        alert = _build_alert(
                            pair=pair, entry_tf=entry_tf,
                            close_time=close_time, c=c, s=s,
                            is_short=is_short, atr_e=atr_e, cfg=cfg, mode=mode,
                            standing=standing,
                        )
                        if alert is not None:
                            emit(s, "RETEST", c,
                                 f"return to origin zone → confirmation entry "
                                 f"@ {c.close:g}", c.close)
                            alerts.append(alert)
                        active[d] = None
                        continue
                if i - s.bos_index > cfg.retest_window:
                    kill(s, "EXPIRED", c, "no retest of the origin zone")
                    continue

    return alerts, events


def _build_alert(
    *,
    pair, entry_tf, close_time, c, s: Setup, is_short, atr_e, cfg, mode, standing
) -> Alert | None:
    entry = c.close
    buf = cfg.sl_buffer_atr * atr_e
    sl = s.inv_level + buf if is_short else s.inv_level - buf
    risk = (sl - entry) if is_short else (entry - sl)
    if risk <= 0 or risk < cfg.min_risk_atr * atr_e:
        return None

    # targets: internal liquidity first, then the nearest external target
    side = "sellside" if is_short else "buyside"
    inner = [
        p.price for p in s.internal_pools
        if p.side == side and ((p.price < entry) if is_short else (p.price > entry))
    ]
    tp1 = (max(inner) if is_short else min(inner)) if inner else None
    tp2 = s.nearest_external_target
    if tp1 == tp2:
        tp2 = None
    if tp1 is None:
        tp1, tp2 = tp2, None
    if tp1 is None:
        return None
    rr1 = abs(tp1 - entry) / risk
    if rr1 < cfg.min_tp_r and tp2 is not None:
        # internal target too close to be useful — target the external draw
        tp1, tp2 = tp2, None
        rr1 = abs(tp1 - entry) / risk
    if rr1 < cfg.min_tp_r:
        return None

    sess = None
    if cfg.sessions_allowlist:
        sess = F.session_for(close_time, cfg.sessions_allowlist)
    status = "PAPER" if mode == "paper" else "SENT"
    suppress_reason = None
    if cfg.sessions_allowlist and sess is None:
        status, suppress_reason = "SUPPRESSED", "outside session allowlist"

    lo = min(entry, tp1)
    hi = max(entry, tp1)
    intermediate = [
        imb.as_dict() for imb in s.imbalances
        if not (imb.hi < lo or imb.lo > hi)
    ]

    return Alert(
        setup_id=s.setup_id, pair=pair, entry_tf=entry_tf, map_tf=cfg.map_tf_label,
        direction=s.direction, entry=entry, stop_loss=sl,
        tp_internal=tp1, tp_external=tp2, candle_close_time=close_time,
        environment=s.environment, phase=s.phase, htf_alignment=s.htf_alignment,
        origin_key_level=s.level.origin_price,
        key_level_type=s.level.kind,
        key_level_bounds=(s.level.zone_lo, s.level.zone_hi),
        key_level_tested=s.level.touches > 0,
        key_level_flipped=s.level.flipped,
        imbalance_context=[imb.as_dict() for imb in s.imbalances],
        internal_liquidity=[p.as_dict() for p in s.internal_pools],
        external_liquidity=[p.as_dict() for p in s.external_pools],
        draw_on_liquidity=s.draw_on_liquidity,
        nearest_external_target=s.nearest_external_target,
        intermediate_zones=intermediate,
        opposing_liquidity_standing=standing,
        sweep_time=s.sweep_time, bos_time=s.bos_time, return_time=c.time,
        invalidation_level=s.inv_level,
        parameter_version=PARAM_VERSION,
        alert_status=status, suppress_reason=suppress_reason,
        session=sess, atr_entry=atr_e, rr_internal=round(rr1, 2),
    )

File: worker/src/engine.ts
/** Layer 2 — XYZ execution engine (confirmation-entry mode). TS port of
 *  slk_bot/slk/engine.py. Replays the trailing `setupWindow` of entry-TF
 *  candles against point-in-time storyline snapshots:
 *
 *    MAP → TOUCH → SWEEP → SHIFT → RETEST → ALERT | INVALID | EXPIRED
 *
 *  Stateless across scans; idempotency comes from the DB (unique setup ids
 *  and unique event keys). Every transition is emitted as an EngineEvent. */
import * as F from "./features";
import { PARAM_VERSION } from "./config";
import { MAP_TF_SECONDS } from "./storyline";
import type {
  Alert, Candle, Direction, EngineEvent, Setup, Storyline,
} from "./types";
import type { StrategyConfig } from "./config";

export interface ScanEntryArgs {
  pair: string;
  entryTf: string;
  tfSeconds: number;
  candles: Candle[];
  snaps: [number, Storyline][];
  cfg: StrategyConfig;
  mode: "paper" | "live";
  provider: string;
}

function setupId(provider: string, pair: string, entryTf: string, d: Direction, level: { kind: string; originPrice: number; originTime: number }): string {
  // deterministic composite id (readable in logs/audit trail)
  return `${provider}:${pair}:${entryTf}:${d}:${level.kind}:${level.originPrice.toFixed(6)}:${new Date(level.originTime).toISOString()}`;
}

function bisectRight(keys: number[], x: number): number {
  let lo = 0;
  let hi = keys.length;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (keys[mid] <= x) lo = mid + 1;
    else hi = mid;
  }
  return lo;
}

export function scanEntry(args: ScanEntryArgs): { alerts: Alert[]; events: EngineEvent[] } {
  const { pair, entryTf, tfSeconds, candles, snaps, cfg, mode, provider } = args;
  const alerts: Alert[] = [];
  const events: EngineEvent[] = [];
  if (candles.length < cfg.pivotLeft + cfg.pivotRight + 6) return { alerts, events };

  // snapshot validity starts when that H4 candle has closed
  const validFrom = snaps.map(([t]) => t + MAP_TF_SECONDS * 1000).sort((a, b) => a - b);
  const storyByKey = new Map<number, Storyline>();
  for (const [t, s] of snaps) storyByKey.set(t + MAP_TF_SECONDS * 1000, s);
  const storyAt = (closeTime: number): Storyline | null => {
    const i = bisectRight(validFrom, closeTime) - 1;
    return i >= 0 ? storyByKey.get(validFrom[i]) ?? null : null;
  };

  const atrE = F.atr(candles, cfg.atrPeriod);
  const [highs, lows] = F.findSwings(candles, cfg.pivotLeft, cfg.pivotRight);
  const conf = cfg.pivotRight;
  const start = Math.max(0, candles.length - cfg.setupWindow);
  const active: { LONG: Setup | null; SHORT: Setup | null } = { LONG: null, SHORT: null };

  const emit = (s: Setup, state: string, c: Candle, reason: string, price: number | null = c.c) => {
    events.push({ setupId: s.setupId, pair, state, candleTime: c.t, reason, price });
    console.info(JSON.stringify({ level: "info", msg: "slk.transition", pair, tf: entryTf, setupId: s.setupId, state, reason, price }));
  };
  const kill = (s: Setup, state: string, c: Candle, reason: string) => {
    emit(s, state, c, reason, c.c);
    active[s.direction] = null;
  };

  for (let i = start; i < candles.length; i++) {
    const c = candles[i];
    const closeTime = c.t + tfSeconds * 1000;
    const story = storyAt(closeTime);

    for (const d of ["SHORT", "LONG"] as Direction[]) {
      const isShort = d === "SHORT";
      let s = active[d];

      // ---- MAP: arm an origin level -----------------------------------
      if (!s && story && story.valid && story.direction === d && story.origin) {
        const lv = story.origin;
        const onSide = isShort ? c.c < lv.zoneLo : c.c > lv.zoneHi;
        if (onSide) {
          s = {
            setupId: setupId(provider, pair, entryTf, d, lv),
            direction: d, level: lv, state: "MAP",
            mapIndex: i, mapTime: c.t,
            touchIndex: 0, touchTime: null,
            sweptPoolIndex: -1, sweptPoolPrice: 0,
            sweepIndex: 0, sweepTime: null,
            extreme: 0, refPrice: 0,
            bosIndex: 0, bosTime: null,
            invLevel: 0, leftZone: false,
            environment: story.environment, phase: story.phase,
            htfAlignment: story.htfAlignment,
            drawOnLiquidity: story.drawOnLiquidity,
            nearestExternalTarget: story.nearestExternalTarget,
            internalPools: [...story.internalPools],
            externalPools: [...story.externalPools],
            imbalances: [...story.imbalances],
          };
          active[d] = s;
          emit(s, "MAP", c, `${lv.kind}-level ${lv.originPrice} armed (${story.environment}/${story.phase})`, lv.originPrice);
        }
      }
      if (!s) continue;
      const z = s.level;
      const cur: Setup = s;

      // ---- HTF story flip ---------------------------------------------
      if (story && story.valid && story.direction !== null && story.direction !== d) {
        kill(cur, "INVALID", c, "HTF storyline invalidation");
        continue;
      }

      const brokeLevel = isShort
        ? c.c > z.zoneHi + cfg.flipMarginAtr * atrE
        : c.c < z.zoneLo - cfg.flipMarginAtr * atrE;

      // ---- MAP → TOUCH -------------------------------------------------
      if (cur.state === "MAP") {
        if (brokeLevel) { kill(cur, "INVALID", c, "close beyond key level (level may flip)"); continue; }
        // Expectation re-evaluation: while still pre-touch, follow the
        // storyline if it moves to a different (fresher/better) origin.
        // Without this, a stale MAP armed from an old snapshot wedges the
        // direction slot and starves newer valid setups. Dedupe-safe:
        // events are unique on (setup_id, state, candle_time).
        const origin = story && story.valid && story.direction === d ? story.origin : null;
        if (origin && (origin.kind !== cur.level.kind || origin.originTime !== cur.level.originTime)) {
          const onSideNew = isShort ? c.c < origin.zoneLo : c.c > origin.zoneHi;
          if (onSideNew) {
            cur.level = origin;
            cur.setupId = setupId(provider, pair, entryTf, d, origin);
            cur.mapIndex = i;
            cur.mapTime = c.t;
            emit(cur, "MAP", c, `${origin.kind}-level ${origin.originPrice} armed (${story!.environment}/${story!.phase})`, origin.originPrice);
          }
        }
        const touched = isShort ? c.h >= z.zoneLo : c.l <= z.zoneHi;
        if (touched) {
          cur.state = "TOUCH";
          cur.touchIndex = i;
          cur.touchTime = c.t;
          emit(cur, "TOUCH", c, "price entered the origin zone");
        } else if (i - cur.mapIndex > cfg.touchWindow) {
          kill(cur, "EXPIRED", c, "level not reached in time");
          continue;
        }
      }

      // ---- TOUCH → SWEEP -----------------------------------------------
      if (cur.state === "TOUCH") {
        if (brokeLevel) { kill(cur, "INVALID", c, "close beyond key level without a sweep"); continue; }
        const pools = isShort ? highs : lows;
        const swept = pools.filter((sw) =>
          sw.index <= cur.touchIndex && sw.index + conf <= i
          && (isShort ? c.h > sw.price && c.c < sw.price : c.l < sw.price && c.c > sw.price)
          && (isShort ? c.h >= z.zoneLo : c.l <= z.zoneHi));
        if (swept.length) {
          const pick = isShort
            ? swept.reduce((a, b) => (b.price > a.price ? b : a))
            : swept.reduce((a, b) => (b.price < a.price ? b : a));
          cur.sweptPoolIndex = pick.index;
          cur.sweptPoolPrice = pick.price;
          cur.sweepIndex = i;
          cur.sweepTime = c.t;
          cur.extreme = isShort ? c.h : c.l;
          const refSwings = (isShort ? lows : highs).filter(
            (rw) => pick.index < rw.index && rw.index <= i && rw.index + conf <= i);
          if (refSwings.length) {
            cur.refPrice = refSwings[refSwings.length - 1].price;
          } else {
            const seg = candles.slice(cur.touchIndex, i + 1);
            cur.refPrice = isShort
              ? Math.min(...seg.map((x) => x.l))
              : Math.max(...seg.map((x) => x.h));
          }
          cur.state = "SHIFT";
          emit(cur, "SWEEP", c, `swept ${isShort ? "buyside" : "sellside"} internal liquidity @ ${pick.price}`, pick.price);
        } else if (i - cur.touchIndex > cfg.sweepWindow) {
          kill(cur, "EXPIRED", c, "no liquidity sweep after the touch");
          continue;
        }
      }

      // ---- SWEEP → SHIFT (BOS) ------------------------------------------
      if (cur.state === "SHIFT") {
        if (i > cur.sweepIndex) {
          const prev = candles[i - 1];
          cur.extreme = isShort ? Math.max(cur.extreme, prev.h) : Math.min(cur.extreme, prev.l);
        }
        const violated = isShort ? c.c > cur.extreme : c.c < cur.extreme;
        if (violated) { kill(cur, "INVALID", c, "close beyond sweep extreme"); continue; }
        const bos = isShort ? c.c < cur.refPrice : c.c > cur.refPrice;
        if (bos) {
          cur.invLevel = isShort ? Math.max(cur.extreme, c.h) : Math.min(cur.extreme, c.l);
          cur.bosIndex = i;
          cur.bosTime = c.t;
          cur.state = "RETEST";
          emit(cur, "SHIFT", c, `BOS through pullback structure ${cur.refPrice}`);
        }
        if (cur.state === "SHIFT" && i - cur.sweepIndex > cfg.bosWindow) {
          kill(cur, "EXPIRED", c, "no BOS after the sweep");
          continue;
        }
      }

      // ---- SHIFT → RETEST → ALERT ---------------------------------------
      if (cur.state === "RETEST") {
        const violated = isShort ? c.c > cur.invLevel : c.c < cur.invLevel;
        if (violated) { kill(cur, "INVALID", c, "close beyond invalidation level"); continue; }
        if (i > cur.bosIndex) {
          const left = isShort ? c.c < z.zoneLo : c.c > z.zoneHi;
          if (left) cur.leftZone = true;
          const tol = cfg.retestToleranceAtr * atrE;
          const returns = isShort ? c.h >= z.zoneLo - tol : c.l <= z.zoneHi + tol;
          if (cur.leftZone && returns) {
            // opposing liquidity must remain standing for reversal setups
            let standing = false;
            if (cur.drawOnLiquidity !== null) {
              const seg = candles.slice(cur.mapIndex, i + 1);
              standing = isShort
                ? seg.every((x) => x.l > (cur.drawOnLiquidity as number))
                : seg.every((x) => x.h < (cur.drawOnLiquidity as number));
            }
            const alert = buildAlert({
              pair, entryTf, closeTime, c, s: cur, isShort,
              atrE, cfg, mode, standing, provider,
            });
            if (alert) {
              emit(cur, "RETEST", c, `return to origin zone → confirmation entry @ ${c.c}`);
              alerts.push(alert);
            }
            active[d] = null;
            continue;
          }
        }
        if (i - cur.bosIndex > cfg.retestWindow) {
          kill(cur, "EXPIRED", c, "no retest of the origin zone");
          continue;
        }
      }
    }
  }

  return { alerts, events };
}

interface BuildAlertArgs {
  pair: string; entryTf: string; closeTime: number; c: Candle; s: Setup;
  isShort: boolean; atrE: number; cfg: StrategyConfig;
  mode: "paper" | "live"; standing: boolean; provider: string;
}

/** Pick the internal-liquidity target (tp1) and external drawback (tp2) for
 *  a setup. Exported for tests. Direction-sanity: the external draw is picked
 *  at STORYLINE-map time and price can run past it before the retest entry
 *  fires — a target on the wrong side of the entry is meaningless, so it is
 *  dropped rather than stored (production: a SHORT XAUUSD alert shipped with
 *  tp above entry and later "resolved" as TP_HIT at -1.2R). */
export function selectTargets(args: {
  isShort: boolean; entry: number; risk: number; minTpR: number;
  internalPools: { side: string; price: number }[];
  nearestExternalTarget: number | null;
}): { tp1: number; tp2: number | null } | null {
  const side = args.isShort ? "sellside" : "buyside";
  const inner = args.internalPools
    .filter((p) => p.side === side && (args.isShort ? p.price < args.entry : p.price > args.entry))
    .map((p) => p.price);
  let tp1: number | null = inner.length
    ? (args.isShort ? Math.max(...inner) : Math.min(...inner))
    : null;
  let tp2 = args.nearestExternalTarget;
  if (tp2 !== null && (args.isShort ? tp2 >= args.entry : tp2 <= args.entry)) tp2 = null;
  if (tp1 === tp2) tp2 = null;
  if (tp1 === null) { tp1 = tp2; tp2 = null; }
  if (tp1 === null) return null;
  let rr1 = Math.abs(tp1 - args.entry) / args.risk;
  if (rr1 < args.minTpR && tp2 !== null) {
    // internal target too close — target the external draw directly
    tp1 = tp2;
    tp2 = null;
    rr1 = Math.abs(tp1 - args.entry) / args.risk;
  }
  if (rr1 < args.minTpR) return null;
  return { tp1, tp2 };
}

function buildAlert(a: BuildAlertArgs): Alert | null {
  const { pair, entryTf, closeTime, c, s, isShort, atrE, cfg, mode, standing, provider } = a;
  const entry = c.c;
  const buf = cfg.slBufferAtr * atrE;
  const sl = isShort ? s.invLevel + buf : s.invLevel - buf;
  const risk = isShort ? sl - entry : entry - sl;
  if (risk <= 0 || risk < cfg.minRiskAtr * atrE) return null;

  // targets: internal liquidity first, then the nearest external target
  const targets = selectTargets({
    isShort, entry, risk, minTpR: cfg.minTpR,
    internalPools: s.internalPools, nearestExternalTarget: s.nearestExternalTarget,
  });
  if (!targets) return null;
  const { tp1, tp2 } = targets;

  let sess: string | null = null;
  if (cfg.sessionsAllowlist.length) sess = F.sessionFor(closeTime, cfg.sessionsAllowlist);
  let status: Alert["alertStatus"] = mode === "paper" ? "PAPER" : "SENT";
  let suppressReason: string | null = null;
  if (cfg.sessionsAllowlist.length && sess === null) {
    status = "SUPPRESSED";
    suppressReason = "outside session allowlist";
  }

  const lo = Math.min(entry, tp1);
  const hi = Math.max(entry, tp1);
  const intermediate = s.imbalances
    .filter((imb) => !(imb.hi < lo || imb.lo > hi))
    .map((imb) => ({ lo: imb.lo, hi: imb.hi, direction: imb.direction }));

  return {
    setupId: s.setupId, pair, entryTf, mapTf: cfg.mapTfLabel,
    direction: s.direction, entry, stopLoss: sl,
    tpInternal: tp1, tpExternal: tp2, candleCloseTime: closeTime,
    environment: s.environment, phase: s.phase, htfAlignment: s.htfAlignment,
    originKeyLevel: s.level.originPrice, keyLevelType: s.level.kind,
    keyLevelBounds: [s.level.zoneLo, s.level.zoneHi],
    keyLevelTested: s.level.touches > 0, keyLevelFlipped: s.level.flipped,
    imbalanceContext: s.imbalances.map((i2) => ({ lo: i2.lo, hi: i2.hi, direction: i2.direction })),
    internalLiquidity: s.internalPools.map((p) => ({ price: p.price, side: p.side, kind: p.kind, time: new Date(p.sourceTime).toISOString() })),
    externalLiquidity: s.externalPools.map((p) => ({ price: p.price, side: p.side, kind: p.kind, time: new Date(p.sourceTime).toISOString() })),
    drawOnLiquidity: s.drawOnLiquidity,
    nearestExternalTarget: s.nearestExternalTarget,
    intermediateZones: intermediate,
    opposingLiquidityStanding: standing,
    sweepTime: s.sweepTime ?? 0, bosTime: s.bosTime ?? 0, returnTime: c.t,
    invalidationLevel: s.invLevel, invalidationReason: null,
    parameterVersion: PARAM_VERSION,
    alertStatus: status, suppressReason, session: sess,
    atrEntry: atrE, rrInternal: Math.round((Math.abs(tp1 - entry) / risk) * 100) / 100,
    cycleStage: "entry_alert", entryMode: "confirmation",
  };
}
